soup/man/man8/prep.8
.\" section 8 man page for prep
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2023.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec mis-features and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH prep.sh 8 "05 January 2023" "prep.sh" "IOCCC tools"
.SH NAME
prep.sh \- prepare for a release - actions for make prep and make release
.SH SYNOPSIS
\fBprep.sh\fP [\-h] [\-v level] [\-V] [\-e] [\-o] [\-m make] [\-M Makefile] [\-l logfile]
.SH DESCRIPTION
\fBprep.sh\fP runs a series of make actions to test the state of any changes prior to making a commit and pull request.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-v\fP
Set verbosity level (def: 0).
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB\-e\fP
Exit on first make action error.
.TP
\fB\-o\fP
Do not use backup files for flex or bison if either of those tools cannot be used.
Specifically if this option is used the make rule \fBparser\-o\fP is used instead of \fBparser\fP.
.TP
\fB\-m \fImake\fP\fP
Specify path to \fBmake(1)\fP command.
.TP
\fB\-M \fIMakefile\fP\fP
Path to the Makefile to use.
.TP
\fB\-l \fIlogfile\fP\fP
Write all details to a log file and write less information to stdout.
.SH EXIT STATUS
.TP
0
all tests passed
.TQ
1
\fB\-h\fP and help string printed or \fB\-V\fP and version string printed
.TQ
2
command line error
.TQ
3
Makefile not readable file that exists
.TQ
4
could not make writable log file
.TQ
>= 10
some make action exited non-zero
.SH NOTES
.PP
There are two ways that this script should be invoked in order of preference:
.TP
.B make release
run the script with the \fB\-e\fP and \fB\-o\fP options.
.TQ
.B make prep
run the script without exiting on the first error and allowing flex or bison backup files
.PP
Although the \fBmake release\fP should be used first it is okay to use \fBmake prep\fP to get a summary of all issues to fix.
However the \fBmake release\fP generally should pass before a commit (if one does not modify the jparse files then one can probably get away with not using \fBmake release\fP).
.PP
This script is only needed for the repo maintainers prior to a commit and pull request.
.PP
.SH EXAMPLES
.PP
.nf
Run the script in a way that aborts on the first error including the inability to use \fBbison\fP or \fBflex\fP.
.RS
\fB
 make release\fP
.RE
.PP
.nf
Run the script in a more lenient way, specifying a log file to reduce output on stdout:
.RS
\fB
 make prep\fP
.RE
.SH SEE ALSO
\fBbison(1)\fP, \fBflex(1)\fP, \fBmake(1)\fP


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 1.0.0 2023-01-06

XXX - add more details here as more is done or remembered (what has already been done).


## Release 0.9.9 2022-12-31

Release just prior to version 1.0: the "public comment version".

The `make prep` and `make release` rules were made much less verbose.

Updated and fixed .gitignore file.

The jparse/ directory only uses and depends on dbg/ and dyn_alloc/.

Made changes to version.h so that the JSON parser code under
jparse/ does not have to use it.

Tuned the verbosity of Makefiles with various Makefile variables such as
${Q}, ${V}, ${S}, ${M}, ${INSTALL_Q} ${INSTALL_V}, and ${MAKE_CD_Q}.

Moved all man pages under their respective sub-directories.  The build_man
rule create a man sub-directory with copies of all other man pages.

Numerous updates and fixes to the man page set.  Improved how man page
macros are used.

Added man symlinks in place where commands are in a given man page.

Change IOCCC use of twitter to mastodon.  The mkiocccentry tool now asks for
an optional mastodon account.  The .auth.json file contains entries for
an optional mastodon instead of twitter.

The Makefile set, and the way they call each other has been improved.
Programs link to one or more libraries such as dbg.a, dyn+array.a,
jparse.a and soup.a.

Make the Makefile set reasonably consistent in terms of style and comments.

Removed use of man2html.  No longer are html files made from man pages.

Removed the checknr rule form Makefiles.  Instead, the check_man rule
tests man pages.

Fix the system of multiple Makefiles (more work might need to be done like
installing man pages with `make install` as well as other possible issues -
several have been discovered and fixed since this issue was 'closed'). Other
miscellany fixes have been made in the Makefiles.

Improve `bug_report.sh`: with `-l` (only write to log file) it is completely
silent and with `-t` fewer tests are run (don't invoke any of the make checks)
making it faster (note though if our programs are not compiled it will cause
some tests to fail: using via `make prep` will solve this problem). The `-x`
option will cause it to delete the report if no issues are found.

Improve `prep.sh`: make it much quieter if `-l logfile` (write to log file and
write less information to stdout) is used. Now calls `bug_report.sh` with
options `-t -x -l` via Makefile rule `bug_report-txl`. This allows for the
`bug_report.sh -t` to work without causing any problems. For reporting bugs one
should call `make bug_report`: `prep.sh` is for the maintainers of the repo.

The JSON test files used to test the chkentry tool were moved to the
test_ioccc/test_JSON directory.

Updated copyright to include 2023 (and a 'Happy New Year' to one and all!).

Add separate man pages for dbg(3) functions. Improve several of the man pages
more generally. TODO: make all man pages installed via `make install`.

Moved jparse to subdirectory.

Removed jparse.md and dbg.md in favor or README.md files in sub-directories.

Fixed dyn_array bug.

Moved dyn_array code to dyn_array/ subdirectory.

Fixed obscure bug with bool typedef.


## Release 0.8.1 2022-11-19

This is "roll-up" release was added so that it could be referenced
in the https://www.ioccc.org/index.html news section.

Updated legacy_clobber rule to remove older directories no longer used.

Moved test directory to test_ioccc to avoid Makefile confusion
with the test rule.

Restructure dbg facility and Makefile in accordance with
the new method of copying code from external GitHub repositories.

Improve formatting with the hostchk_warning rule.

Added initial dyn_array.md documentation.

Typo fixes to the verge.8 man page.

Minor fixes to jparse.md.

Updated txzchk comments (source, header, man page).

Improved and updated README.md documentation.

Improved bug_report.sh: in particular in reporting versions.

Bug fixes and improvements to several shell scripts.
Made changes to satisfy shellcheck as well.

Added rules in the Makefile to make is easier to import code from
external GitHub repositories.

```
This repo is designed to be a standalone repo.  Even though we use other
repositories, we prefer to NOT clone them.  We want this repo to depend
on a specific version of such code so that a change in the code of those
external repositories will NOT impact this repo.

For that reason, and others, we maintain a private copy of an external
repository as clone.repo.  The clone.repo directory is excluded from
out repo via the .gitignore file.  We copy clone.repo/ into repo/ and
check in those file directly into this repo.
```

The dbg code is now a copy of the code from the dbg repo.

Update JSON parser version.  Changed JSON_PARSER_VERSION from "0.11
2022-11-04" to "0.12 2022-11-09".  There was an API change (an
addition of a filename argument in some calls) that suggested version change.


## Release 0.8 2022-11-09

Improve the situation with some systems requiring some feature test macros.
Removed all references to `${TIMEGM_PROBLEM}` in Makefile.
Removed all references to `${D_LEGACY}` in Makefile.
Cleaned out the `$(WARN_FLAGS}` list (from legacy hosts).
These changes allow for cleanly compiling under CentOS without as many
complexities (it was also important for modern systems like Fedora).

Made a number of important fixes to `chkentry`.

Made a number of important fixes to JSON parser.
Pre-scan documents and strings for problematic bytes prior to trying to parse it
as json.

The JSON scanner and parser are now re-entrant!

Improved JSON parser error messages by adding filename and location (lines and
columns).

Added filename argument to top level parser functions:

```c
extern struct json *parse_json(char const *ptr, size_t len, char const *filename, bool *is_valid);
extern struct json *parse_json_stream(FILE *stream, char const *filename, bool *is_valid);
```

This was one part of making the scanner and parser re-entrant. See the git log

test_ioccc/prep.sh
#!/usr/bin/env bash
#
# prep.sh - prep for a release - actions for make prep and make release
#
# This script was written in 2022 by:
#
#	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# with some minor improvements by:
#
#	@xexyl
#	https://xexyl.net		Cody Boone Ferguson
#	https://ioccc.xexyl.net
#
# "Because sometimes even the IOCCC Judges need some help." :-)
#
# Share and enjoy! :-)


# setup
#
export FAILURE_SUMMARY=
export LOGFILE=
export PREP_VERSION="0.2 2023-01-04"
export USAGE="usage: $0 [-h] [-v level] [-V] [-e] [-o] [-m make] [-M Makefile] [-l logfile]

    -h              print help and exit
    -v level        flag ignored
    -V              print version and exit

    -e		    exit on first make action error (def: exit only at end)
    -o		    do NOT use backup files, fail if bison or flex cannot be used (def: use)
    -m make	    make command (def: make)
    -M Makefile	    path to Makefile (def: ./Makefile)
    -l logfile      write details of actions to logfile (def: send to stdout)

Exit codes:
     0   all tests passed
     1   -h and help string printed or -V and version string printed
     2	 command line error
     3	 Makefile not a readable file that exists
     4	 could not make writable log file
 >= 10   some make action exited non-zero

$0 version: $PREP_VERSION"
export MAKE="make"
export MAKEFILE="./Makefile"
export V_FLAG="0"
export E_FLAG=
export EXIT_CODE="0"
export O_FLAG=

# parse args
#
while getopts :hv:Veom:M:l: flag; do
    case "$flag" in
    h)	echo "$USAGE" 1>&2
	exit 1
	;;
    v)	V_FLAG="$OPTARG";
	;;
    V)	echo "$PREP_VERSION"
	exit 1
	;;
    m)	MAKE="$OPTARG";
	;;
    M)	MAKEFILE="$OPTARG";
	;;
    e)	E_FLAG="-e"
	;;
    o)	O_FLAG="-o"
	;;
    l)	LOGFILE="$OPTARG"
	;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
	echo 1>&2
	echo "$USAGE" 1>&2
	exit 2
	;;
    :)	echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
	echo 1>&2
	echo "$USAGE" 1>&2
	exit 2
	;;
    *)
	;;
    esac
done

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -ne 0 ]]; then
    echo "$0: ERROR: expected 0 arguments, found $#" 1>&2
    exit 2
fi

# firewall
#
if [[ ! -e $MAKEFILE ]]; then
    echo "$0: ERROR: Makefile not found: $MAKEFILE" 1>&2
    exit 3
fi
if [[ ! -f $MAKEFILE ]]; then
    echo "$0: ERROR: Makefile not a regular file: $MAKEFILE" 1>&2
    exit 3
fi
if [[ ! -r $MAKEFILE ]]; then
    echo "$0: ERROR: Makefile not a readable file: $MAKEFILE" 1>&2
    exit 3
fi

# if -l logfile was specified, remove it and recreate it to start out empty
#
if [[ -n "$LOGFILE" ]]; then
    rm -f "$LOGFILE"
    touch "$LOGFILE"
    if [[ ! -f "${LOGFILE}" ]]; then
	echo "$0: ERROR: couldn't create log file" 1>&2
	exit 4
    fi
    if [[ ! -w "${LOGFILE}" ]]; then
	echo "$0: ERROR: log file not writable" 1>&2
	exit 4
    fi
fi

# write_echo - write a message to either the log file or both the log file and
# stdout
#
write_echo()
{
    local MSG="$*"

    if [[ -n "$LOGFILE" ]]; then
	echo "$MSG" | tee -a -- "$LOGFILE"
    else
	echo "$MSG" 1>&2
    fi
}

# write_echo_n - write a message to either the log file or both the log file and
# stdout but without a trailing newline
write_echo_n()
{
    local MSG="$*"

    if [[ -n "$LOGFILE" ]]; then
	echo -n "$MSG" | tee -a -- "$LOGFILE"
    else
	echo -n "$MSG" 1>&2
    fi
}

# exec_command - invoke command redirecting output only to the log file or to
# both stdout and the log file
exec_command()
{
    local COMMAND=$*
    if [[ -n "$LOGFILE" ]]; then
	# prep.sh:169:10: note: Double quote to prevent globbing and word splitting. [SC2086]
	#
	# shellcheck disable=SC2086
	command ${COMMAND} >> "$LOGFILE" 2>&1
	return $?
    else
	# prep.sh:169:10: note: Double quote to prevent globbing and word splitting. [SC2086]
	#
	# shellcheck disable=SC2086
	command ${COMMAND} 2>&1
	return $?
    fi

}

# make action
#
# usage:
#	make_action code rule
#
#	code - exit code of rule fails
#	rule - Makefile rule to call
#
make_action() {

    # parse args
    #
    if [[ $# -ne 2 ]]; then
	echo "$0: ERROR: function expects 2 args, found $#" 1>&2
	exit 9
    fi
    local CODE="$1"
    local RULE="$2"

    # announce pre-action
    #
    if [[ -z "$LOGFILE" ]]; then
	write_echo "=-=-= START: $MAKE $RULE =-=-="
	write_echo "$MAKE" -f "$MAKEFILE" "$RULE"
    else
	write_echo_n "make_action $CODE $RULE "
    fi

    # perform action
    #
    exec_command "$MAKE" -f "$MAKEFILE" "$RULE"
    status="$?"
    if [[ $status -ne 0 ]]; then

	# process a make action failure
	#
	EXIT_CODE="$CODE"

	FAILURE_SUMMARY="$FAILURE_SUMMARY
	make_action $EXIT_CODE: $MAKE -f $MAKEFILE $RULE: non-zero exit code: $status"
	if [[ -z "$LOGFILE" ]]; then
	    write_echo "$0: Warning: EXIT_CODE is now: $EXIT_CODE" 1>&2
	fi
	if [[ -n $E_FLAG ]]; then
	    if [[ -z "$LOGFILE" ]]; then
		write_echo
		write_echo "$0: ERROR: $MAKE -f $MAKEFILE $RULE exit status: $status" 1>&2
		write_echo
		write_echo "=-=-= FAIL: $MAKE $RULE =-=-="
		write_echo
	    else
		write_echo "ERROR exit code $status"
	    fi
	    exit "$EXIT_CODE"
	else
	    if [[ -z "$LOGFILE" ]]; then
		write_echo
		write_echo "$0: Warning: $MAKE -f $MAKEFILE $RULE exit status: $status" 1>&2
		write_echo
		write_echo "=-=-= FAIL: $MAKE $RULE =-=-="
		write_echo
	    else
		write_echo "ERROR exit code $status"
	    fi
	fi

    # announce post-action
    #
    else
	if [[ -z "$LOGFILE" ]]; then
	    write_echo
	    write_echo "=-=-= PASS: $MAKE $RULE =-=-="
	    write_echo
	else
	    write_echo "OK"
	fi
    fi
    return 0;
}

# perform make actions
#
if [[ -z "$LOGFILE" ]]; then
    write_echo "=-=-=-=-= START: $0 =-=-=-=-="
    write_echo
fi
make_action 10 clobber
make_action 11 all
make_action 12 depend
make_action 13 clean_mkchk_sem
make_action 14 all_sem_ref
make_action 15 mkchk_sem
make_action 16 all
if [[ -n $O_FLAG ]]; then
    make_action 17 parser-o
else
    make_action 18 parser
fi
make_action 19 all
make_action 20 load_json_ref
make_action 21 use_json_ref
make_action 22 clean_generated_obj
make_action 23 all
make_action 24 bug_report-txl
make_action 25 shellcheck
make_action 26 seqcexit
make_action 27 picky
make_action 28 tags
make_action 29 check_man
make_action 30 all
make_action 31 test

if [[ $EXIT_CODE -eq 0 ]]; then
    if [[ -z "$LOGFILE" ]]; then
	write_echo "=-=-=-=-= PASS: $0 =-=-=-=-="
	write_echo
    else
	write_echo "All tests PASSED."
	write_echo ""
	write_echo "See test_ioccc/test_ioccc.log for more details."
    fi
else
    if [[ -z "$LOGFILE" ]]; then
	write_echo "=-=-=-=-= FAIL: $0 =-=-=-=-="
	write_echo
	write_echo "=-=-=-=-= Will exit: $EXIT_CODE =-=-=-=-="
	write_echo
	if [[ -n "$FAILURE_SUMMARY" ]]; then
	    write_echo "One or more tests failed:"
	    write_echo "$FAILURE_SUMMARY"
	    write_echo ""
	fi
	write_echo "See test_ioccc/test_ioccc.log for more details."
    else
	if [[ -n "$FAILURE_SUMMARY" ]]; then
	    write_echo "One or more tests failed:"
	    write_echo "$FAILURE_SUMMARY"
	    write_echo ""
	else
	    write_echo "One or more tests failed."
	fi
	write_echo "See test_ioccc/test_ioccc.log for more details."
    fi
fi

# All Done!!! -- Jessica Noll, Age 2
#
exit "$EXIT_CODE"
